### DISCUSSION TITLE
Incremental generator / caching

### DISCUSSION THREAD

**Discussion by mnaoumov** [srid/emanote#343] [Ideas]
For large sites, it will be a huge performance impact to regenerate only changed files and skip rebuilding unchanged ones

**Comment by srid** [comment]
This already happens in live-server (the `run` command). Are you talking about static site generation (the `gen` command)?

I never considered optimizing it for the static site generation case since we usually use live-server for the rapid-feedback cycle and use static generation in a one-off manner.

**Reply by srid** [reply]
Sometimes I keep the live-server running as a daemon and use it to view my notes as I edit them over time in text editors.

See "Using Emanote as a Home Manager service" in https://emanote.srid.ca/tips/nix

**Reply by mnaoumov** [reply]
Yes, I am talking about `gen` command. The optimization there is the must for making efficient site updates

**Reply by mnaoumov** [reply]
@srid Home Manager service scenario won't fit in my case. I use `Obsidian` on the daily basis, so I don't need to use live server daemon.

I was looking for a good way to export my blog from `Obsidian` to publish online and `emanote` was the best solution I could find.

It was lacking one critical to me feature so I had to create a PR for that (#341), hopefully you will make it into the main build soon.

Anyway, bringing caching to the `gen` command will improve the tool heavily for scenarios like mine (blog updates)

**Reply by srid** [reply]
How big is your blog? How long does `gen` take currently? How often do you run `gen`?

**Reply by mnaoumov** [reply]
My blog has

1927 md files = 1.6 Mb

222 attachments: pdf/images = 68 Mb

I run them on WSL2 on Windows 11

`gen` takes ~40 mins.

I can't say how often because today I used it the first time. But I think I prefer to update my blog regularly, when I write a new post, I want to publish it straight away

**Reply by srid** [reply]
2000 files is not a lot but 40 mins is extremely slow. Perhaps we can optimize it. Could profile Haskell code to see where most slowness is.

**Reply by mnaoumov** [reply]
@srid I am a software developer myself, but I don't have an experience with Haskell. You can guide me how to troubleshoot the slowness. Maybe it's related to my environment. I am not sure how performant WSL2 is. But the performance is a different issue anyway. We can benefit from the incremental generator anyway

**Reply by srid** [reply]
You can try `cabal build --enable-profiling --ghc-options="-fno-prof-auto"` in the nix shell (`nix develop`). This blog post may be useful: https://www.tweag.io/blog/2020-01-30-haskell-profiling/
